File: 1325-path-with-maximum-probability/1325-path-with-maximum-probability-BELLMAN-FORD.java
class Solution {
    public double maxProbability(int n, int[][] edges, double[] succProb, int start_node, int end_node) {
        double[] cost=new double[n];
        cost[start_node]=1;
        for(int i=0;i<n-1;i++){
            boolean changes=false;
            for(int j=0;j<edges.length;j++){
                int u=edges[j][0],v=edges[j][1];
                double edgeCost=succProb[j];
                if(edgeCost*cost[u]>cost[v]){
                    cost[v]=edgeCost*cost[u];
                    changes=true;
                }
                if(edgeCost*cost[v]>cost[u]){
                    cost[u]=edgeCost*cost[v];
                    changes=true;
                }
            }
            if(!changes){
                break;
            }
        }
        return cost[end_node];
    }
}

File: 1325-path-with-maximum-probability/1325-path-with-maximum-probability-BELLMAN-FORD.py
#Belmann-ford algorithm
class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start_node: int, end_node: int) -> float:
        cost=[0]*n
        cost[start_node]=1
        for i in range(n-1):
            changes=False
            for j in range(len(edges)):
                u,v=edges[j]
                edgeCost=succProb[j]

                #u-->v
                if edgeCost*cost[u]>cost[v]:
                    cost[v]=edgeCost*cost[u]
                    changes=True
                #v-->u
                if edgeCost*cost[v]>cost[u]:
                    cost[u]=edgeCost*cost[v]
                    changes=True
            if not changes:
                break
        return cost[end_node]
                
